### PR TITLE
[apex] New rule ApexUnitTestClassShouldHaveRunAs #4149

### LINKED ISSUE #4149: [apex] new rule : ApexUnitTestClassShouldHaveRunAs
**Proposed Rule Name: ApexUnitTestClassShouldHaveRunAs**

**Proposed Category:**  [Best Practices ]

**Description:**
According to Salesforce best practices, when executing test method we should not rely on the user's running context.
The use of [System.runAs](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_testing_tools_runas.htm) is recommended to execute the test with a newly created user so that the test does not rely on environment data and user's configuration

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
//Code to be flagged by the rule
@isTest
static void testNoRunAs() {
      Profile p = [SELECT Id FROM Profile WHERE Name='Standard User'];
      User u2 = new User(Alias = 'newUser', Email='newuser@testorg.com',
         EmailEncodingKey='UTF-8', LastName='Testing', LanguageLocaleKey='en_US',
         LocaleSidKey='en_US', ProfileId = p.Id,
         TimeZoneSidKey='America/Los_Angeles', UserName='newuser@testorg.com');

      System.debug('Current User: ' + UserInfo.getUserName());
   }
   
//Code OK
@isTest
private class TestRunAs {

   public static testMethod void testIgnoreCase() {

      Profile p = [SELECT Id FROM Profile WHERE Name='Standard User'];
      User usr = new User(Alias = 'newUser', Email='newuser@testorg.com',
         EmailEncodingKey='UTF-8', LastName='Testing', LanguageLocaleKey='en_US',
         LocaleSidKey='en_US', ProfileId = p.Id,
         TimeZoneSidKey='America/Los_Angeles', UserName='newuser@testorg.com');

      System.runAs(usr) {
         System.debug('Current User: ' + UserInfo.getUserName());
      }
   }
}
```

### PR DESCRIPTION
**Pull request by tprouvot**
## Describe the PR

According to Salesforce best practices, when executing test method we should not rely on the user's running context.
The use of [System.runAs](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_testing_tools_runas.htm) is recommended to execute the test with a newly created user so that the test does not rely on environment data and user's configuration

## Related issues
- Fixes #4149

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by tprouvot** [comment]
Hi @adangel !
I may have missed some steps, it's the first time I create a new rule.
Feel free to reach me if you want any updates on the PR.

File: pmd-apex/src/main/resources/rulesets/apex/apexunit.xml [hunk 1]
```diff
@@ -9,6 +9,7 @@
     </description>
 
     <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveAsserts" deprecated="true" />
+    <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveRunAs" deprecated="true" />
```
Comment (adangel) [inline pmd-apex/src/main/resources/rulesets/apex/apexunit.xml:12]: This ruleset is deprecated anyway, we don't need to maintain it, so we can leave this file untouched.

https://pmd.github.io/latest/pmd_rules_apex.html#additional-rulesets

> Deprecated This ruleset is for backwards compatibility.

File: pmd-apex/src/main/resources/rulesets/apex/quickstart.xml [hunk 2]
```diff
@@ -250,6 +250,9 @@
          <property name="cc_block_highlighting" value="false" />
       </properties>
    </rule>
+   <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveRunAs" message="Apex unit test classes should have at least one System.runAs() call">
+      <priority>3</priority>
+   </rule>
```
Comment (adangel) [inline pmd-apex/src/main/resources/rulesets/apex/quickstart.xml:255]: Just use a simpe reference without overriding any value. I don't know, why the messages for the other rule refs have been overridden... it is unnecessary, unless you really want a different message...
Also the priority is inherited from the original rule definition.

```suggestion
   <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveRunAs" />
```

Actually the messages _are_ different...

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml [hunk 3]
```diff
@@ -64,6 +64,44 @@ public class Foo {
         </example>
     </rule>
 
+    <rule name="ApexUnitTestClassShouldHaveRunAs"
+          language="apex"
+          since="6.51.0"
+          message="Apex unit tests should use System.runAs() method"
```
Comment (adangel) [inline pmd-apex/src/main/resources/category/apex/bestpractices.xml:70]:
```suggestion
          message="Apex unit test classes should have at least one System.runAs() call"
```

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for your PR!

Please have a look at my comments.

Currently the rule basically only checks, whether in the test method at least once `System.runAs` has been used. Is this sufficient? It's a very simple check... (but [ApexUnitTestClassShouldHaveAsserts](https://pmd.github.io/latest/pmd_rules_apex_bestpractices.html#apexunittestclassshouldhaveasserts) does only check for at least one assert as well).

Maybe this simple check is enough for the beginning, maybe not - I don't know. That would add up to false-negatives - e.g. Statements that are outside of a RunAs block, but should actually have been inside a RunAs block.

And it will really flag all tests, that don't use System.runAs - maybe there are cases where you don't need a user, since you are testing just some plain logic. That would add up to false-positives.

I'm not an Apex developer, so I can't tell.

The documentation on salesforce also doesn't have a real example, because you won't test a System.debug call as they do in a RunAs... and there are no asserts in the examples as well...

File: pmd-apex/src/main/resources/rulesets/apex/quickstart.xml (hunk 2 above)
Comment (tprouvot) [reply pmd-apex/src/main/resources/rulesets/apex/quickstart.xml:255]: > 

You're right, initially I did not call the constructor in the java class to define the message and priority.
Thanks !

**Comment by tprouvot** [comment]
> Maybe this simple check is enough for the beginning, maybe not - I don't know. That would add up to false-negatives - e.g. Statements that are outside of a RunAs block, but should actu

Hi @adangel thanks for your review ! 
As you said this rule is pretty simple and a kind of copy of [ApexUnitTestClassShouldHaveAsserts](https://pmd.github.io/latest/pmd_rules_apex_bestpractices.html#apexunittestclassshouldhaveasserts).

As you mentioned it, it may exist some cases where we don't need a user to run the test but most of the time we do.
We could say that we don't need runAs if we do not perform CRUD operation in the test method but it is hard to know when we call external methods which is frequently used to create test data..

I think this version is a good start and we will see if we need to extend the way it works.

File: pmd-core/src/main/resources/rulesets/releases/6510.xml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+<ruleset name="6510"
+    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
+  <description>
+This ruleset contains links to rules that are new in PMD v6.51.0
+  </description>
+
+    <rule ref="rulesets/apex/apexunit.xml/ApexUnitTestClassShouldHaveRunAs" />
```
Comment (adangel) [inline pmd-core/src/main/resources/rulesets/releases/6510.xml:11]:
```suggestion
    <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveRunAs" />
```

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 141 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/d956e05f4e204b71b69a9b34ab2240dc/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**